ops/charm.py

```diff
@@ -1297,16 +1335,48 @@ def __init__(self, raw: Optional[Dict[str, Any]] = None,
                          for name, storage in raw_.get('storage', {}).items()}
         self.resources = {name: ResourceMeta(name, res)
                           for name, res in raw_.get('resources', {}).items()}
+        self.devices = {name: DeviceMeta.from_dict(device)
+                        for name, device in raw_.get('devices', {}).items()}
         self.payloads = {name: PayloadMeta(name, payload)
                          for name, payload in raw_.get('payloads', {}).items()}
         self.extra_bindings = raw_.get('extra-bindings', {})
         self.actions = {name: ActionMeta(name, action) for name, action in actions_raw_.items()}
-        # This is taken from Charm Metadata v2, but only the "containers" and
-        # "containers.name" fields that we need right now for Pebble. See:
-        # https://discourse.charmhub.io/t/charm-metadata-v2/3674
         self.containers = {name: ContainerMeta(name, container)
                            for name, container in raw_.get('containers', {}).items()}
 
+    def _load_links(self, raw: Dict[str, Any]):
+        websites = raw.get('website', [])
+        if not websites and 'links' in raw:
+            # When running tests, this might be loading from charmcraft.yaml.
```

Comment:
What about just having this comment once at the top of the function?

Reply:
Good call, will do.